File: Taller/TallerEnPython/JumpSearch.py
import math
import time
import json
import csv
import os
import random

# --- Implementación del Algoritmo Jump Search ---

def jump_search(arr, x):
    """
    Algoritmo de búsqueda por saltos.
    Fuente: https://www.geeksforgeeks.org/dsa/jump-search/
    """
    n = len(arr)
    # Tamaño del bloque a saltar: raíz cuadrada de n
    step = int(math.floor(math.sqrt(n)))
    prev = 0

    # Encontrar el bloque donde el elemento está presente
    # Mientras el valor en el paso actual sea menor que x
    while arr[min(step, n) - 1] < x:
        prev = step
        step += int(math.floor(math.sqrt(n)))
        if prev >= n:
            return -1

    # Búsqueda lineal en el bloque identificado
    while arr[prev] < x:
        prev += 1
        # Si llegamos al siguiente bloque o al final, no está
        if prev == min(step, n):
            return -1

    # Si se encuentra el elemento
    if arr[prev] == x:
        return prev

    return -1

# --- Funciones de Utilidad (Simulando ArchivoUtils) ---

def leer_archivo(nombre_archivo):
    if not os.path.exists(nombre_archivo):
        print(f"Error: El archivo {nombre_archivo} no existe.")
        return []
    with open(nombre_archivo, 'r') as f:
        return [int(linea.strip()) for linea in f if linea.strip()]

def generar_numero_aleatorio():
    return random.randint(10000000, 99999999)

resultados_registro = []

def probar_archivo(nombre_archivo, numero, algoritmo):
    print(f"\nProcesando: {nombre_archivo}")
    
    datos = leer_archivo(nombre_archivo)
    if not datos:
        return

    # Jump Search requiere un arreglo ordenado
    datos.sort()

    # Medición en nanosegundos (equivalente a System.nanoTime)
    inicio = time.perf_counter_ns()
    resultado = jump_search(datos, numero)
    fin = time.perf_counter_ns()

    tiempo_ns = fin - inicio
    print(f"Tiempo (ns): {tiempo_ns}")

    if resultado != -1:
        print(f"Número encontrado en posición: {resultado}")
    else:
        print("Número NO encontrado")

    resultados_registro.append({
        "algoritmo": algoritmo,
        "archivo": nombre_archivo,
        "tiempo_ns": tiempo_ns
    })

def exportar_datos(algoritmo):
    # CSV
    with open(f"{algoritmo}.csv", 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=["algoritmo", "archivo", "tiempo_ns"])
        writer.writeheader()
        writer.writerows(resultados_registro)
    
    # JSON
    with open(f"{algoritmo}.json", 'w') as f:
        json.dump(resultados_registro, f, indent=4)
    print(f"\nResultados exportados a {algoritmo}.csv y {algoritmo}.json")

# --- Bloque Principal ---

if __name__ == "__main__":
    nombre_algoritmo = "jumpsearch"
    
    # Generar el número de 8 dígitos a buscar
    numero_buscado = generar_numero_aleatorio()
    print(f"Número a buscar: {numero_buscado}")

    # Pruebas con los diferentes archivos
    probar_archivo("archivo_10k.txt", numero_buscado, nombre_algoritmo)
    probar_archivo("archivo_100k.txt", numero_buscado, nombre_algoritmo)
    probar_archivo("archivo_1M.txt", numero_buscado, nombre_algoritmo)

    # Exportar resultados finales
    exportar_datos(nombre_algoritmo)

File: Taller/TallerEnJava/src/main/java/com/analisis/algoritmos/JumpSearch.java
package com.analisis.algoritmos;

public class JumpSearch {

public static void probarArchivo(String nombreArchivo, int numero, String algoritmo) {

    System.out.println("\nProcesando: " + nombreArchivo);

    int[] datos = ArchivoUtils.leerArchivo(nombreArchivo);

    DualPivotQuicksort.dualPivotQuickSort(datos, 0, datos.length - 1);

    int resultado = -1;

    int repeticiones = 100;

    for (int i = 0; i < 50; i++) {
        jumpSearch(datos, numero);
    }

    long inicio = System.nanoTime();

    for (int i = 0; i < repeticiones; i++) {
        resultado = jumpSearch(datos, numero);
    }

    long fin = System.nanoTime();

    long tiempoPromedioNs = (fin - inicio) / repeticiones;

    System.out.println("Tiempo promedio (ns): " + tiempoPromedioNs);

    if (resultado != -1) {
        System.out.println("Número encontrado en posición: " + resultado);
    } else {
        System.out.println("Número NO encontrado");
    }

    ArchivoUtils.agregarResultadoNs(algoritmo, nombreArchivo, tiempoPromedioNs);
}

    public static void main(String[] args) {

        String algoritmo = "jumpsearch";

        ArchivoUtils.limpiarResultados();

        int numeroBuscado = BinarySearch.generarNumeroAleatorio();
        System.out.println("Número a buscar: " + numeroBuscado);

        probarArchivo("archivo_10k.txt", numeroBuscado, algoritmo);
        probarArchivo("archivo_100k.txt", numeroBuscado, algoritmo);
        probarArchivo("archivo_1M.txt", numeroBuscado, algoritmo);

        ArchivoUtils.exportarCSVNs(algoritmo);
        ArchivoUtils.exportarJSONNs(algoritmo);
    }
    
    /*
    * Fuente: https://www.geeksforgeeks.org/dsa/jump-search/
     */
    public static int jumpSearch(int[] arr, int x){
        int n = arr.length;

        // Finding block size to be jumped
        int step = (int)Math.floor(Math.sqrt(n));

        // Finding the block where element is
        // present (if it is present)
        int prev = 0;
        for (int minStep = Math.min(step, n)-1; arr[minStep] < x; minStep = Math.min(step, n)-1){
            prev = step;
            step += (int)Math.floor(Math.sqrt(n));
            if (prev >= n)
                return -1;
        }

        // Doing a linear search for x in block
        // beginning with prev.
        while (arr[prev] < x){
            prev++;

            // If we reached next block or end of
            // array, element is not present.
            if (prev == Math.min(step, n))
                return -1;
        }

        // If element is found
        if (arr[prev] == x)
            return prev;

        return -1;
    }
}
